### DISCUSSION TITLE
Does LazyVim have C/C++ autocompletion/goto symbol enabled by default?

### DISCUSSION THREAD

**Discussion by jerinjoy** [LazyVim/LazyVim#571] [Q&A]
I've installed LazyVim and am editing some C++ code. C++ autocomplete doesn't seem to be enabled - typing a Class object name followed by a `.` doesn't bring up a prompt with the functions in the class.

Also, `gd` and `<leader>sS` when the cursor is on a function name doesn't go to the function definition and instead shows:

```
[telescope.builtin.lsp_*]: no client attached
```

**Comment by dpetka2001** [comment]
There's no LSP server installed by default. You should install (maybe clangd?) or whatever you like to have such features.

**Reply by jerinjoy** [reply]
thanks, I installed `clangd` and `clang-format` with mason.

That fixed the issue partially.
I `nvim` open in in a C++ project directory. `<leader>sS` still prints:

```
[telescope.builtin.lsp_*]: no client attached
```

If I open a file, `<leader>ss` will show the symbols in the open buffer.

Is there anything else I need to do to get the LSP to parse all the files in my directory?

I also see errors about standard C types:
```
Unknown type name `uint32_t`
```

**Reply by dpetka2001** [reply]
I think for clangd you need to have a compile_commands.json at the root of your project for the LSP server to kick in. Please see [here](https://sarcasm.github.io/notes/dev/compilation-database.html) for more info. A quick command to generate it for a single file would be `clang-12 -MJ compile_commands.json -Wall -Werror -o main -c main.c` I guess? For more complicated projects you should look at CMake maybe?

**Reply by amaanq** [reply]
clangd doesn't need compile_commands.json

**Reply by dpetka2001** [reply]
Is that so? Was under the impression it needed it. But just tried it out without the file and clangd kicks in just fine. From your experience when is it needed?

**Reply by dpetka2001** [reply]
This is the opposite of what's stated in the lspconfig [docs](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#clangd). If you don't mind I would welcome an explanation about this.

**Reply by amaanq** [reply]
> Is that so? Was under the impression it needed it. But just tried it out without the file and clangd kicks in just fine. From your experience when is it needed?

not sure, any c/c++ file triggers it for me

**Reply by dpetka2001** [reply]
Yes I can see that's the case after you mentioned it. Just perplexed about the reference from lspconfig docs.

**Reply by jerinjoy** [reply]
@amaanq you're saying it should "just work" if the `clangd` LSP is installed through `mason`?

**Reply by dpetka2001** [reply]
Yes, it should. I also didn't have clangd installed. Just installed it from mason and when I open a test.c file the client attaches correctly to the buffer. Check what `:LspInfo` tells you. Mine is telling me that it runs in single file mode when i tested with test.c single file only.

**Reply by jerinjoy** [reply]
The `:LspInfo` output looks ok:

```
 Language client log: /Users/joy/.local/state/nvim/lsp.log
 Detected filetype:   cpp
 
 1 client(s) attached to this buffer: 
 
 Client: clangd (id: 1, bufnr: [19])
 	filetypes:       c, cpp, objc, objcpp, cuda, proto
 	autostart:       true
 	root directory:  /Users/joy/workspace/cppproj
 	cmd:             clangd
 
 Configured servers list: pylsp, clangd, lua_ls, jsonls
```

but `<leader>sS` only shows symbols from the current buffer and it still complains about standard types being unknown:
```
Unknown type name `uint32_t`
```

**Reply by dpetka2001** [reply]
Not sure what to tell you. I don't code in C++, so I'm not familiar. `<leader>sS` in a test file that I tried quickly shows no errors for me. But that was just a basic hello_world program. Maybe someone more experienced in C++ will be able to help you out.

**Reply by amaanq** [reply]
uint32_t isn't a default type, it requires including stdint.h or cstdint for C and C++ respectively. That's normal.

**Reply by jerinjoy** [reply]
thanks, is the LSP limited to just the open files?
there are other headers in the root directory that include the headers for `uint32_t`.
The global symbol search doesn't show any symbols that aren't in the current open file.

**Reply by amaanq** [reply]
I don't believe so, it'll be fine as long as any header you include in your current file has uint32 defined/included somewhere.

**Reply by GauthierLi** [reply]
it decide by your complile tools, if you use cmake or bazel to complie your project, it will have different way to generate a **compile.json stuff file ,wihch could be solved by clangd. after that you could use auto completion or jump function.
